overlord/configstate/configcore/journal.go

```diff
@@ -0,0 +1,92 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/sys"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/systemd"
+)
+
+var osutilFindGid = osutil.FindGid
+var sysChownPath = sys.ChownPath
+
+func init() {
+	supportedConfigurations["core.journal.persistent"] = true
+}
+
+func validateJournalSettings(tr config.ConfGetter) error {
+	return validateBoolFlag(tr, "journal.persistent")
+}
+
+func handleJournalConfiguration(tr config.ConfGetter, opts *fsOnlyContext) error {
+	output, err := coreCfg(tr, "journal.persistent")
+	if err != nil {
+		return nil
+	}
+
+	if output == "" {
+		return nil
+	}
+
+	var sysd systemd.Systemd
+	rootDir := dirs.GlobalRootDir
+	if opts != nil {
+		rootDir = opts.RootDir
+	}
+
+	logPath := filepath.Join(rootDir, "/var/log/journal")
+
+	switch output {
+	case "true":
+		if err := os.MkdirAll(logPath, 0755); err != nil {
+			return err
+		}
+		gid, err := osutilFindGid("systemd-journal")
+		if err != nil {
+			return fmt.Errorf("cannot find systemd-journal group: %v", err)
+		}
+		if err := sysChownPath(logPath, 0, sys.GroupID(gid)); err != nil {
+			return err
+		}
+	case "false":
+		if err := os.RemoveAll(logPath); err != nil {
```

Comment:
This looks fine. I'm a bit worried though about preexisting systems where people manipulated directly their journald.conf and/or created the dir by themselves, but there's no easy way to dump journald current config. Maybe when we create the dir we should live a marker `.snapd-created`  if the dir didn't exist already and check it here and error instead of removing if the marker is not there, mentioning that the dir was setup outside of snapd control.

Reply:
We could remove only the file we create and remove the directory only if it is empty.

Reply:
to implement that we probably need to make /var/log/journal-snapd/.snapd-created first
and then rename /var/log/journal-snapd to /var/log/journal, so that we don't create the directory without the marker.  The /var/log/journal-snapd if it's around we probably clean it any time we reach this code.